src/chi-vue/src/components/data-table/DataTable.tsx

```diff
@@ -0,0 +1,852 @@
+            );
+          }
+        });
+      }
+    });
+    (this.$refs.dataTable as HTMLElement).addEventListener('mouseup', () => {
+
+      setTimeout(() => {
+        this._preventSortOnResize = false;
+      }, 0);
+      thElm = null;
+    });
+  }
+
+  cellAlignment(align: string): string {
+    const alignmentUtilityClasses: any = {
+      left: UTILITY_CLASSES.JUSTIFY.START,
+      center: UTILITY_CLASSES.JUSTIFY.CENTER,
+      right: UTILITY_CLASSES.JUSTIFY.END
+    };
+
+    if (align) {
+      return alignmentUtilityClasses[align];
+    }
+
+    return '';
+  }
+
+  dataTableClasses(style: DataTableStyleConfig, sortable: boolean) {
+    return [
+      DATA_TABLE_CLASSES.DATA_TABLE,
+      style.portal ? DATA_TABLE_CLASSES.PORTAL : '',
+      style.bordered ? DATA_TABLE_CLASSES.BORDERED : '',
+      style.noBorder ? DATA_TABLE_CLASSES.NO_BORDER : '',
+      style.hover ? DATA_TABLE_CLASSES.HOVER : '',
+      style.striped ? DATA_TABLE_CLASSES.STRIPED : '',
+      style.size ? `-${style.size}` : '',
+      sortable ? DATA_TABLE_CLASSES.SORTING : ''
+    ].join(' ');
+  }
+
+  expansionButton(rowId: string) {
+    const expansionIcons: DataTableExpansionIcons = {
+      portal: {
+        expanded: 'minus',
+        collapsed: 'plus'
+      },
+      base: {
+        expanded: 'chevron-up',
+        collapsed: 'chevron-right'
+      }
+    };
+    const iconStyle: DataTableExpansionIconStyles = this.config.style
+      .portal
+      ? 'portal'
+      : 'base';
+
+    return (
+      <div
+        class={`${DATA_TABLE_CLASSES.CELL} ${DATA_TABLE_CLASSES.EXPANDABLE}`}
+        role="cell"
+      >
+        <button
+          class={`
+        ${BUTTON_CLASSES.BUTTON}
+        ${BUTTON_CLASSES.ICON_BUTTON}
+        ${BUTTON_CLASSES.FLAT}
+        -sm
+        `}
+          aria-label="Expand row"
+          data-target={`#row-${rowId}-content`}
+          onClick={() => this.toggleRow(rowId.toString())}
+        >
+          <div class={BUTTON_CLASSES.CONTENT}>
+            <i
+              class={`
+            ${ICON_CLASS}
+            icon-${
+              this.accordionsShown.includes(rowId)
+                ? expansionIcons[iconStyle].expanded
+                : expansionIcons[iconStyle].collapsed
+            }
+            `}
+            />
+          </div>
+          <span class={SR_ONLY}>Expand Row</span>
+        </button>
+      </div>
+    );
+  }
+
+  selectRow(rowId: string) {
+    const indexOfRowId = this.selectedRows.indexOf(rowId);
+
+    if (indexOfRowId === -1) {
+      this.selectedRows.push(rowId);
+    } else {
+      this.selectedRows.splice(indexOfRowId, 1);
+    }
+    this.$emit(DATA_TABLE_EVENTS.SELECTED_ROWS_CHANGE, this.selectedRows);
+  }
+
+  selectAllRows(action: 'select' | 'deselect') {
+    if (action === 'select') {
+      this.slicedData.forEach((row: DataTableRow) => {
+        if (row.id && !this.selectedRows.includes(`row-${this._dataTableNumber}-${row.id}`)) {
```

Comment:
Here to manipulate select all/de-selectall row.id value is used, does it provides flexibility to use this feature in case if the row does not has 'id' attribute in the table data?.

Can we parametrize this key from table column configurations?.

Reply:
Good catch! I changed the approach.
In order to emit row selection event with row corresponding data, I added formatting of received data (`serializeRow` function in `created` lifecycle hook) to add a unique `rowNumber` to each row.

The `rowNumber`s are fixed and will not be mutated if the Data Table is re-rendered. This way row data will always be accessible.

Reply:
So, a row on 4th level would have an id:
row-dt-0-5-1-1-1

`dt-0` - is the Data Table number
`5` - 6th element on the first level
`1` - 2nd element on second level
`1` - 2nd lement on third level
`1` - 2nds element on fourth level

Here is an example of the final markup
![Screenshot 2021-02-07 at 20 35 11](https://user-images.githubusercontent.com/25136131/107157396-37efe400-6984-11eb-9a1d-5f3d1b9ee736.png)

Reply:
@devi-das @dani-cl-madrid